nipst/post.go

```diff
@@ -21,78 +19,74 @@ func DefaultConfig() config.Config {
 	validator := validation.NewValidator(&cfg)
-	err := validator.Validate((*proving.Proof)(proof))
-	if err != nil {
-		return false, err
+	if err := validator.Validate((*proving.Proof)(proof)); err != nil {
+		return err
 	}
 
-	return true, nil
+	return nil
 }
 
 type PostClient struct {
+	id          []byte
 	cfg         *config.Config
 	initializer *initialization.Initializer
+	prover      *proving.Prover
 	logger      shared.Logger
 }
 
 // A compile time check to ensure that PostClient fully implements PostProverClient.
 var _ PostProverClient = (*PostClient)(nil)
 
-func NewPostClient(cfg *config.Config) *PostClient {
-	return &PostClient{cfg, nil, shared.DisabledLogger{}}
+func NewPostClient(cfg *config.Config, id []byte) *PostClient {
+	return &PostClient{
+		id:          id,
+		cfg:         cfg,
+		initializer: initialization.NewInitializer(cfg, id),
+		prover:      proving.NewProver(cfg, id),
+		logger:      shared.DisabledLogger{},
+	}
 }
 
-func (c *PostClient) initialize(id []byte, timeout time.Duration) (commitment *types.PostProof, err error) {
-	// TODO(moshababo): implement timeout
-	//TODO: implement persistence
-	if c.cfg.SpacePerUnit%merkle.NodeSize != 0 {
-		return nil, fmt.Errorf("space (%d) is not a multiple of merkle.NodeSize (%d)", c.cfg.SpacePerUnit, merkle.NodeSize)
-	}
-	c.initializer = initialization.NewInitializer(c.cfg, id)
+func (c *PostClient) Initialize() (commitment *types.PostProof, err error) {
 	proof, err := c.initializer.Initialize()
 	return (*types.PostProof)(proof), err
 }
 
-func (c *PostClient) execute(id []byte, challenge []byte, timeout time.Duration) (*types.PostProof, error) {
-	// TODO(moshababo): implement timeout
-	prover := proving.NewProver(c.cfg, id)
-	proof, err := prover.GenerateProof(challenge)
+func (c *PostClient) Execute(challenge []byte) (*types.PostProof, error) {
+	proof, err := c.prover.GenerateProof(challenge)
 	return (*types.PostProof)(proof), err
 }
 
 func (c *PostClient) Reset() error {
-	if c.initializer != nil {
-		err := c.initializer.Reset()
-		if err != nil {
-			return err
-		}
-	}
-	return nil
+	return c.initializer.Reset()
 }
 
-func (c *PostClient) SetLogger(logger shared.Logger) {
-	c.logger = logger
+func (c *PostClient) IsInitialized() bool {
+	state, _, _ := c.initializer.State()
+	return state == initialization.StateCompleted
 }
 
 func (c *PostClient) SetParams(dataDir string, space uint64) {
 	cfg := *c.cfg
 	c.cfg = &cfg
 	c.cfg.DataDir = dataDir
 	c.cfg.SpacePerUnit = space
+
+	c.initializer = initialization.NewInitializer(c.cfg, c.id)
+	c.prover = proving.NewProver(c.cfg, c.id)
 }
 
-func (c *PostClient) Initialized() bool {
-	if c.initializer == nil {
-		return false
-	}
+func (c *PostClient) SetLogger(logger shared.Logger) {
+	c.logger = logger
+}
 
-	state, _, _ := c.initializer.State()
-	return state == initialization.StateCompleted
+func (c *PostClient) Cfg() *config.Config {
```

Comment:
To what purpose we have this accessor? is it supposed to allow modifications? If so, ,ultithreading issue should be taken into account, if not, I'd recommend returning a copy

Reply:
The purpose is to remove `postCfg` from from `nipstBuilder`, so we won't have it in 2 places.
In this case, it need to be read from `postClient`. 

So this isn't to allow writes, only reads.
I'll return a copy instead, as you suggested.

Reply:
Fixed.